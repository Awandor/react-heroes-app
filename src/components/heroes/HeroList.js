import { useMemo } from 'react';
import { getHeroesByPublisher } from '../../selectors/getHeroesByPublisher';
import { HeroCard } from './HeroCard';

export const HeroList = ( { publisher } ) => {

    // const heroes = getHeroesByPublisher(publisher);

    const heroes = useMemo( () => getHeroesByPublisher( publisher ), [ publisher ] );

    return (
        <div className="card-columns animate__animated animate__fadeIn">
            {
                heroes.map( hero => (

                    /* Usamos la desestructuraci√≥n de objetos para enviar las props */

                    <HeroCard key={hero.id} {...hero} />
                ) )
            }
        </div>
    );

};
