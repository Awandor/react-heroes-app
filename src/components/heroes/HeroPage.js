import { Redirect, useParams, useHistory } from "react-router-dom";
import { getHeroById } from "../../selectors/getHeroById";
import { useMemo } from 'react';

export const HeroPage = () => {

    // const params = useParams();

    // console.log(params);

    const { heroId } = useParams();

    // const hero = getHeroById(heroId);

    const hero = useMemo( () => getHeroById( heroId ), [ heroId ] );

    // console.log( 'hero: ', hero );

    let history = useHistory();

    // si heroId es undefined la app se rompe al hacer la desestructuración así que validamos

    if ( !hero ) {

        console.log( 'NO EXISTE' );

        return <Redirect to="/" />;

    }

    const { superhero, publisher, alter_ego, first_appearance, characters } = hero;

    const handleReturn = () => {

        // console.log( 'Volver ha sido pulsado ++++++++++++++++++++++++++++' );

        // console.log( 'history.length', history.length );

        if( history.length <= 2 ){

            // console.log( 'PUSH' );

            history.push( '/' );

        } else {

            // console.log( 'GOBACK' );

            history.goBack();

        }

    };

    return (
        <div className="row mt-5">
            <div className="col-4">
                <img src={`../assets/heroes/${heroId}.jpg`} alt="{superhero}" className="img-thumbnail animate__animated animate__fadeInLeft" />
            </div>

            <div className="col-8">
                <h3>{superhero}</h3>
                <ul className="list-group list-group-flush">
                    <li className="list-group-item">
                        <b>Alter ego:</b> {alter_ego}
                    </li>
                    <li className="list-group-item">
                        <b>Publisher:</b> {publisher}
                    </li>
                    <li className="list-group-item">
                        <b>First appearance:</b> {first_appearance}
                    </li>
                </ul>

                <h5>Characters</h5>
                <p>{characters}</p>

                <button className="btn btn-info" onClick={handleReturn}>Volver</button>
            </div>
        </div>
    );

};
