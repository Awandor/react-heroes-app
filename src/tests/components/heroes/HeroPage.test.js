import { mount } from "enzyme";
import { MemoryRouter, Route, Router } from "react-router-dom";
import { HeroPage } from "../../../components/heroes/HeroPage";

describe( 'Pruebas en <HeroPage />', () => {

    // Vamos a hacer un mock del history tal y como lo usamos en HeroPage

    const historyMock = {
        length: 10,
        push: jest.fn(),
        goBack: jest.fn(),

        // Nos fijamos en los errores de lo que falta y los vamos añadiendo deduciendo el tipo
        listen: jest.fn(),
        createHref: jest.fn(),
        replace: jest.fn(),

        location: {
            pathname: '/hero/marvel-spider'
        }

    };

    // Ahora lo aplicamos al HeroPage usando Router

    /* const wrapper = mount(
        <MemoryRouter initialEntries={ [ '/hero' ] }>
            <Router history={historyMock}>
                <HeroPage />
            </Router>
        </MemoryRouter>
    ); */

    // Es conveniente limpiar los mocks después de cada prueba

    afterEach( () => {

        jest.clearAllMocks();

    } );



    test( 'debe mostrar el componente <Redirect /> si no hay parámetros en el URL', () => {

        const wrapperNoHero = mount(
            <MemoryRouter initialEntries={[ '/hero' ]}>
                <HeroPage />
            </MemoryRouter>
        );

        expect( wrapperNoHero.find( 'Redirect' ).exists() ).toBe( true );

    } );

    const wrapper = mount(
        <MemoryRouter initialEntries={[ '/hero/marvel-spider' ]}>
            <Route path="/hero/:heroId">
                <HeroPage />
            </Route>
        </MemoryRouter>
    );


    test( 'debe mostrar un hero si el parámetro es correcto', () => {

        expect( wrapper.find( 'h3' ).text().trim() ).toBe( 'Spider Man' );

    } );

    const wrapperHistory = mount(
        <MemoryRouter initialEntries={[ '/hero/marvel-spider' ]}>
            <Router history={historyMock}>
                <Route path="/hero/:heroId" component={() => <HeroPage />} />
            </Router>
        </MemoryRouter>
    );

    test( 'debe regresar a la pantalla anterior con goBack', () => {

        expect( wrapperHistory ).toMatchSnapshot();

        // Simulamos el click del botón Volver

        wrapperHistory.find( 'button' ).simulate( 'click' );

        // Vamos a evaluar si la función history.goBack dentro de handleReturn ha sido invocada

        expect( historyMock.goBack ).toHaveBeenCalled();

    } );

    test( 'debe ir a la pantalla de inicio', () => {

        historyMock.length = 1;

        // Simulamos el click del botón Volver

        wrapperHistory.find( 'button' ).simulate( 'click' );

        // Vamos a evaluar si la función history.goBack dentro de handleReturn ha sido invocada

        expect( historyMock.push ).toHaveBeenCalledWith( '/' );

    } );

    test( 'debe llamar Redirect si el héroe no existe', () => {

        historyMock.location = {
            pathname: '/hero/algo'
        };

        const wrapperHistory = mount(
            <MemoryRouter initialEntries={[ '/hero/algo' ]}>
                <Router history={historyMock}>
                    <Route path="/hero/:heroId" component={() => <HeroPage />} />
                </Router>
            </MemoryRouter>
        );

        expect( wrapperHistory ).toMatchSnapshot();

    } );


} );
