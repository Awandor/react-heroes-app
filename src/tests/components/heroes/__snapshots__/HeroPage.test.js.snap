// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <HeroPage /> debe llamar Redirect si el h√©roe no existe 1`] = `""`;

exports[`Pruebas en <HeroPage /> debe regresar a la pantalla anterior con goBack 1`] = `
<div
  className="row mt-5"
>
  <div
    className="col-4"
  >
    <img
      alt="{superhero}"
      className="img-thumbnail animate__animated animate__fadeInLeft"
      src="../assets/heroes/marvel-spider.jpg"
    />
  </div>
  <div
    className="col-8"
  >
    <h3>
      Spider Man
    </h3>
    <ul
      className="list-group list-group-flush"
    >
      <li
        className="list-group-item"
      >
        <b>
          Alter ego:
        </b>
         
        Peter Parker
      </li>
      <li
        className="list-group-item"
      >
        <b>
          Publisher:
        </b>
         
        Marvel Comics
      </li>
      <li
        className="list-group-item"
      >
        <b>
          First appearance:
        </b>
         
        Amazing Fantasy #15
      </li>
    </ul>
    <h5>
      Characters
    </h5>
    <p>
      Peter Parker
    </p>
    <button
      className="btn btn-info"
      onClick={[Function]}
    >
      Volver
    </button>
  </div>
</div>
`;
